src/shims/thread.rs

```diff
@@ -0,0 +1,138 @@
+use crate::*;
+use rustc_index::vec::Idx;
+use rustc_target::abi::LayoutOf;
+
+impl<'mir, 'tcx> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
+    fn pthread_create(
+        &mut self,
+        thread: OpTy<'tcx, Tag>,
+        _attr: OpTy<'tcx, Tag>,
+        start_routine: OpTy<'tcx, Tag>,
+        arg: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        this.tcx.sess.warn(
+            "thread support is experimental. \
+             For example, Miri does not detect data races yet.",
+        );
+
+        let new_thread_id = this.create_thread()?;
+        let old_thread_id = this.set_active_thread(new_thread_id)?;
+
+        let thread_info_place = this.deref_operand(thread)?;
+        let thread_info_type = thread.layout.ty
+            .builtin_deref(true)
+            .ok_or_else(|| err_ub_format!(
+                "wrong signature used for `pthread_create`: first argument must be a raw pointer."
+            ))?
+            .ty;
+        let thread_info_layout = this.layout_of(thread_info_type)?;
+        this.write_scalar(
+            Scalar::from_uint(new_thread_id.index() as u128, thread_info_layout.size),
+            thread_info_place.into(),
+        )?;
+
+        let fn_ptr = this.read_scalar(start_routine)?.not_undef()?;
+        let instance = this.memory.get_fn(fn_ptr)?.as_instance()?;
+
+        let func_arg = this.read_immediate(arg)?;
+        let func_args = [*func_arg];
+
+        let ret_place =
+            this.allocate(this.layout_of(this.tcx.types.usize)?, MiriMemoryKind::Machine.into());
+
+        this.call_function(
+            instance,
+            &func_args[..],
+            Some(ret_place.into()),
+            StackPopCleanup::None { cleanup: true },
+        )?;
+
+        this.set_active_thread(old_thread_id)?;
+
+        Ok(0)
+    }
+
+    fn pthread_join(
+        &mut self,
+        thread: OpTy<'tcx, Tag>,
+        retval: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        if !this.is_null(this.read_scalar(retval)?.not_undef()?)? {
+            throw_unsup_format!("Miri supports pthread_join only with retval==NULL");
```

Comment:
In other words, this means we do not support getting the return value of the thread, right?

That doesn't seem too hard, but we can leave it for the future with a FIXME.
```suggestion
            // FIXME: implement reading the thread function's return place.
            throw_unsup_format!("Miri supports pthread_join only with retval==NULL");
```

Reply:
Oh no, the docs say this is about `pthread_exit`. Yeah that's not something I see a good reason to support for now.

But what about the return value of the thread root function? Is that just discarded? We are allocating a return place for it, seems weird that that is never used.

Reply:
> In other words, this means we do not support getting the return value of the thread, right?

Since the standard library [always ignores](https://github.com/rust-lang/rust/blob/ce1ab355c27faa84f984b15800cd6a65e674b9f2/src/libstd/sys/unix/thread.rs#L191) the return value, I did not bother implementing support for it.

> Oh no, the docs say this is about `pthread_exit`. But what about the return value of the thread root function? Is that just discarded?

Passing `X` into `pthread_exit` is the same thing as returning from the thread root function with `return X`. [Documentation](http://man7.org/linux/man-pages/man3/pthread_create.3.html):

    It returns from start_routine().  This is equivalent to calling pthread_exit(3) with the value supplied in the return statement.

Reply:
> Since the standard library always ignores the return value, I did not bother implementing support for it.

Hm... fair enough, but please add a comment along those lines at the code that allocates this return place.